crates/circuit/src/dag_circuit.rs

```diff
@@ -6974,6 +6861,271 @@ impl DAGCircuit {
+            self.dag
+                .dag
+                .add_edge(clbit_last_node, new_node, Wire::Clbit(clbit));
+        }
+
+        // If available, check all the vars in this instruction
+        for var in vars.iter().flatten() {
+            let var_last_node = *self.last_vars[var.index()].get_or_insert_with(|| {
+                // If the var is not in the last nodes collection, the edge between the output node and its predecessor.
+                // Then, store the predecessor's NodeIndex in the last nodes collection.
+                let output_node = self.dag.var_io_map[var.index()][1];
+                let (edge_id, predecessor_node) = self
+                    .dag
+                    .dag
+                    .edges_directed(output_node, Incoming)
+                    .next()
+                    .map(|edge| (edge.id(), edge.source()))
+                    .unwrap();
+                self.dag.dag.remove_edge(edge_id);
+                predecessor_node
+            });
+
+            // Because `DAGCircuit::additional_wires` can return repeated instances of vars,
+            // we need to make sure to skip those to avoid cycles.
+            self.last_vars[var.index()] = Some(new_node);
+            if var_last_node == new_node {
+                continue;
+            }
+            self.dag
+                .dag
+                .add_edge(var_last_node, new_node, Wire::Var(*var));
+        }
+        new_node
+    }
+
+    #[inline]
+    fn pack_instruction(
+        &mut self,
+        op: PackedOperation,
+        qubits: Option<OwnedOrSlice<Qubit>>,
+        clbits: Option<OwnedOrSlice<Clbit>>,
+        params: Option<SmallVec<[Param; 3]>>,
+        extra_attrs: ExtraInstructionAttributes,
+        #[cfg(feature = "cache_pygates")] py_op: Option<PyObject>,
+    ) -> PackedInstruction {
+        #[cfg(feature = "cache_pygates")]
+        let py_op = if let Some(py_op) = py_op {
+            py_op.into()
+        } else {
+            OnceLock::new()
+        };
+        let qubits = if let Some(qubits) = qubits {
+            match qubits {
+                OwnedOrSlice::Owned(owned) => self.dag.qargs_interner.insert_owned(owned),
+                OwnedOrSlice::Slice(slice) => self.dag.qargs_interner.insert(slice),
+            }
+        } else {
+            self.dag.qargs_interner.get_default()
+        };
+        let clbits = if let Some(clbits) = clbits {
+            match clbits {
+                OwnedOrSlice::Owned(owned) => self.dag.cargs_interner.insert_owned(owned),
+                OwnedOrSlice::Slice(slice) => self.dag.cargs_interner.insert(slice),
+            }
+        } else {
+            self.dag.cargs_interner.get_default()
+        };
+        PackedInstruction {
+            op,
+            qubits,
+            clbits,
+            params: params.map(|x| x.into()),
+            extra_attrs,
+            #[cfg(feature = "cache_pygates")]
+            py_op,
+        }
+    }
```

Comment:
While the interners do allow for switching on whether the input is owned or not, I think it's worth checking if there's actually all that much benefit to exposing it in a dynamic manner like this.  We'll be hitting the "not present" path really infrequently over all (the vast majority of insertions are cache hits), so there'll be very few actual allocations, whereas moving things in and out of the `Cow` (or `SliceOrOwned`) has a cost on every call.

I doubt `Cow` will be noticeably slower, but I'd also doubt that it'll be notably faster, and it might not be worth the code complexity at the call sites.

Reply:
These now use the `Interner::insert_cow()` method that you designed :)